### ISSUE TITLE
Adding '.' to env:PATH adds/remembers the directory where the command was issued, does not denote the ambient current directory later on

### ISSUE THREAD

**Issue by cklutz** [PowerShell/PowerShell#15795]
### Prerequisites

- [X] Write a descriptive title.
- [X] Make sure you are able to repro it on the [latest released version](https://github.com/PowerShell/PowerShell/releases)
- [X] Search the existing issues.
- [X] Refer to the [FAQ](https://github.com/PowerShell/PowerShell/blob/master/docs/FAQ.md).
- [X] Refer to [Differences between Windows PowerShell 5.1 and PowerShell](https://docs.microsoft.com/powershell/scripting/whats-new/differences-from-windows-powershell).

### Steps to reproduce

When adding '.' to $env:PATH _inside_ PowerShell (like suggested in #14106), the '.' seems to refer to the directory where the `$env:PATH = $env:PATH + ';.'` was issued, but does not denote the actual "current directory", when you change the current directory (using `Set-Location`).

```
(PS> $env:PATH="")
PS> mkdir $env:TEMP\a
PS> mkdir $env:TEMP\b
PS> echo "@echo from a" > $env:TEMP\a\t.cmd
PS> echo "@echo from b" > $env:TEMP\b\t.cmd
PS> cd $env:TEMP\a
PS> $env:PATH=$env:PATH + ';.'
PS> t.cmd
from a
PS> cd $env:TEMP\b
PS> t.cmd
from a
```

That last line should output `from b` (also see "Expected haviour" below). Note that clearing `env:PATH` before all this is "optional" - the behavior is the same.

Interestingly, if you repeat `$env:PATH=$env:PATH + ';.'` in another directory (like above in `$env:TEMP\b`), then this new directory will be considered for the '.' (that is above invoking `t.cmd` would output `from b`, regardless of where  the actual `Get-Location` is at).

Example:
```
PS> cd $env:TEMP\a
PS> $env:PATH=$env:PATH + ';.'
PS> t.cmd
from a
PS> cd $env:TEMP\b
PS> $env:PATH=$env:PATH + ';.'
PS> t.cmd
from b
PS> cd $env:TEMP\a
PS> t.cmd
from b
```

Tests have been done with 7.1.3, but same behavior can be observed with 7.2.0-preview7.
 

### Expected behavior

```console
PS> cd $env:TEMP\a
PS> $env:PATH=$env:PATH + ';.'
PS> t.cmd
from a
PS> cd $env:TEMP\a
PS> t.cmd
from b
```

### Actual behavior

```console
PS> cd $env:TEMP\a
PS> $env:PATH=$env:PATH + ';.'
PS> t.cmd
from a
PS> cd $env:TEMP\a
PS> t.cmd
from a
```

### Environment data

```powershell
Name                           Value
----                           -----
PSVersion                      7.1.3
PSEdition                      Core
GitCommitId                    7.1.3
OS                             Microsoft Windows 10.0.18363
Platform                       Win32NT
PSCompatibleVersions           {1.0, 2.0, 3.0, 4.0…}
PSRemotingProtocolVersion      2.3
SerializationVersion           1.1.0.1
WSManStackVersion              3.0
```

**Comment by cklutz** [comment]
Hi, any news or opinions on this?

**Comment by timkgithub** [comment]
I vote to get this triaged also.

Adding '.\' to the **PATH** list should always resolve to the current path when you change directories!

**Comment by daxian-dbw** [comment]
It looks to me PowerShell command discovery somehow cached the `t.cmd` it found from the first time. From your example repro above, continue to run the following:
```
> cd c:\   # move to a completely unrelated location, which doesn't contain a 't.cmd' file
> t.cmd
from a
```

**Comment by mklement0** [comment]
@daxian-dbw, it looks like it isn't the _specific command_ whose location is cached, it is _the entire `$env:PATH` value_, with `.` getting resolved to the current directory _at the time of caching_.

Given PowerShell's - sensible - requirement to use `.\` or `./` to invoke executables located in the current directory, I think it would make more sense to simply _ignore_ attempts to  add `.` to `$env:PATH`.

Arguably, even the current - obscure and probably accidental - behavior introduces a security vulnerability from that perspective.

**Comment by cklutz** [comment]
@mklement0  I agree with your general sentiment about security. On some unix systems (or with some accounts) you also don't want to have the dot in your path and invoke all commands explicitly using `./command`. 

*However, that should be the user's choice! Not the default forced upon by the shell.*

(One could image a variable, say, `$CurrentDirectoryWarningBehavior` with a value of `Enabled`, that prints a warning every time PS uses the `.` to resolve a command from the path. - if that is possible at all, I don't know) 

Because this issue is a regression to existing behavior and "prior work", confusing and annoying to users coming from or being experienced with:

- PowerShell (for Windows)
- a _diversity_ of UNIX shells
- CMD.exe

**Comment by mklement0** [comment]
@cklutz, you're right, I didn't realize it was a regression (I mistakenly thought it was broken in Windows PowerShell too). 

By the way, on macOS a `.` entry is seemingly _ignored_ altogether, whereas WSL with Ubuntu 20.04 seems to exhibit the same behavior as on Windows.

We're clearly dealing with a bug that needs to be resolved one way or the other.

I see your point about not forcing a choice upon the user, and personally I have no strong feelings about how this should be resolved.

To mention another possibility: emit an _unconditional_ warning if `.` is found in `$env:PATH` _on startup_ or whenever `$env:PATH` is modified in session with a value that has a `.` entry.

**Comment by mklement0** [comment]
P.S., @cklutz

As a workaround or possibly even something you may adopt as an alternative to putting `.` in `$env:PATH` you can rely on _tab-completion_ to avoid having to type `.\` or `./` every time.

This works as intended by default on Windows, but requires some tweaking on Unix-like platforms for a smooth experience (see [this SO answer](https://stackoverflow.com/a/76000155/45375)).

**Comment by szalapski** [comment]
I gotta agree with [timkgithub](https://github.com/timkgithub).  Let it do the obvious thing that everyone would expect.

**Comment by fguitton** [comment]
This is still very much a problem. A resolution would still be rather welcome. 🌴

**Comment by szalapski** [comment]
Please don't close issues that are real and affecting many.

**Comment by JamesWTruher** [comment]
The @WG group discussed this and agree that it's a bug. While there is a security aspect about this, it's still an explicit action that the user has taken. However, the real issue is that if you add "." to your path, the _wrong_ location has been added to you path, once you change your location the previous location is remembered.
and `$env:PATH` may not represent the actual path (since `.` has been changed to `/tmp` but is not reflected.

**Comment by fguitton** [comment]
This needs to stay open a bit longer; presumably until @JamesWTruher comes back to us with another update on this bug 🐞

**Comment by fguitton** [comment]
The recent release of PowerShell 7.5.0 has now completely broken the expected behaviour. When using a `PATH` with `.\node_modules\.bin\` for local repository access to package "binaries" it has become utterly unfunctional. We need this issue to stay open.

**Comment by cklutz** [comment]
Thanks @fguitton for also chiming in. Indeed, an update on the status of this would be welcome (especially now since @JamesWTruher seems to have retired since end of last year).

**Comment by fguitton** [comment]
Since the ticket is still assigned to @JamesWTruher and by now, many many pages down the list, I will _"shoot in the dark"_ and tag a couple of the Microsoft team members (from the top contributor of the past 6 months) for them to hopefully give a bit of clarity or at least perhaps provide an update.

With my sincere apologises for doing so ... @adityapatwardhan, @TravisEz13, @SteveL-MSFT, would you be so kind to look at this issue and let us know what the future is for the inclusion of dynamic (from whichever current directory) relative paths to the `PATH` environment variable, ever so necessary for ecosystems like Node and its `./node_modules/bin` directory.

Please don't bit 🙏🏼 😅 
Thank you very much for your support !

**Comment by s-jacob-powell** [comment]
Just ran into this issue after switching my full workflow with WSL to use PowerShell and windows.
I was surprised it didn't already do this (expected behavior).

**Comment by fguitton** [comment]
@microsoft-github-policy-service
This is still an ongoing issue/request. 🚜 
Anyone having an update about this ?
Have a beautiful day all !

**Comment by cklutz** [comment]
@microsoft-github-policy-service Please keep this alive. There hasn't been any proper or conclusive reaction yet.

**Comment by cklutz** [comment]
Please reopen this issue or at least explain why it was closed. (Can't be six months of inactivity, counting from my last message on July 22nd).
